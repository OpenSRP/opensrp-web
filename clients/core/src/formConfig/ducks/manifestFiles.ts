import {
    fetchActionCreatorFactory,
    removeActionCreatorFactory,
    reducerFactory,
    getItemsByIdFactory,
    getItemsArrayFactory,
    getItemByIdFactory,
} from '../../store/ducks/baseDux';

/** manifest files interface */
export interface ManifestFilesTypes {
    created_at: string;
    id: string;
    identifier: string;
    is_draft: boolean;
    is_json_validator: boolean;
    jursdiction: string;
    label: string;
    module: string;
    version: string;
}

/** reducer name */
export const reducerName = 'manifestFiles';

/** manifest files Reducer */
const reducer = reducerFactory<ManifestFilesTypes>(reducerName);

// action
/** fetch manifest files to store action */
export const fetchManifestFiles = fetchActionCreatorFactory<ManifestFilesTypes>(reducerName, 'id');

/** clear manifest files data in store action*/
export const removeManifestFiles = removeActionCreatorFactory(reducerName);

// selectors
export const getAllManifestFilesById = getItemsByIdFactory<ManifestFilesTypes>(reducerName);
export const getManifestFilesById = getItemByIdFactory<ManifestFilesTypes>(reducerName);
export const getAllManifestFilesArray = getItemsArrayFactory<ManifestFilesTypes>(reducerName);

export default reducer;
