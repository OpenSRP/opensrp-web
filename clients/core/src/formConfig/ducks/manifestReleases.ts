import {
    fetchActionCreatorFactory,
    removeActionCreatorFactory,
    reducerFactory,
    getItemsByIdFactory,
    getItemsArrayFactory,
    getItemByIdFactory,
    getTotalRecordsFactory,
} from '../../store/ducks/baseDux';

/** json field interface */
export interface ManifestJsonFieldType {
    forms_version: string;
    identifiers: string[];
}

/** manifest releses interface */
export interface ManifestReleasesTypes {
    appId: string;
    appVersion: string;
    createdAt: string;
    identifier: string;
    json: string;
    updatedAt: string;
}

/** reducer name */
export const reducerName = 'manifestReleases';

/** manifest releases Reducer */
const reducer = reducerFactory<ManifestReleasesTypes>(reducerName);

// action
/** fetch manifest releases to store action */
export const fetchManifestReleases = fetchActionCreatorFactory<ManifestReleasesTypes>(reducerName, 'identifier');

/** clear manifest releases data in store action*/
export const removeManifestReleases = removeActionCreatorFactory(reducerName);

// selectors
export const getAllManifestReleasesById = getItemsByIdFactory<ManifestReleasesTypes>(reducerName);
export const getManifestReleasesById = getItemByIdFactory<ManifestReleasesTypes>(reducerName);
export const getAllManifestReleasesArray = getItemsArrayFactory<ManifestReleasesTypes>(reducerName);
export const getTotalManifestReleasesRecords = getTotalRecordsFactory(reducerName);

export default reducer;
