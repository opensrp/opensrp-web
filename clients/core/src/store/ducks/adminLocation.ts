import {
    fetchActionCreatorFactory,
    removeActionCreatorFactory,
    setTotalRecordsFactory,
    reducerFactory,
    getItemsByIdFactory,
    getItemsArrayFactory,
    getItemByIdFactory,
    getTotalRecordsFactory,
} from './baseDux';

/** The reducer name */
export const reducerName = 'opensrp-web/admin/location';

/** Interface for Location object */

export interface LocationTag {
    id: number;
    active: boolean;
    name: string;
    description: string;
}
export interface Location {
    type: string;
    id: number | string;
    properties: {
        uid: string;
        type: string;
        status: string;
        parentId: string;
        name: string;
        geographicLevel: number;
        effectiveStartDate: string;
        version: number;
        parentName: string;
        tagName: string;
    };
    serverVersion: number;
    locationTags: LocationTag[];
}

/** Client Reducer */
const reducer = reducerFactory<Location>(reducerName);

// action creators
export const fetchLocationList = fetchActionCreatorFactory<Location>(reducerName, 'id');
export const removeLocationList = removeActionCreatorFactory(reducerName);
export const setTotalRecords = setTotalRecordsFactory(reducerName);

// selectors
export const getLocationListById = getItemsByIdFactory<Location>(reducerName);
export const getLocationArray = getItemsArrayFactory<Location>(reducerName);
export const getLocationById = getItemByIdFactory<Location>(reducerName);
export const getTotalRecords = getTotalRecordsFactory(reducerName);

export default reducer;
