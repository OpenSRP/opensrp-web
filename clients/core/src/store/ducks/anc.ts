import {
    reducerFactory,
    BaseClient,
    fetchClientsFactory,
    removeClientsFactory,
    getClientsArrayFactory,
    getClientByIdFactory,
    getClientsByIdFactory,
} from './baseClient';
import { FlexObject } from '../../helpers/utils';

/** describes an ANC client object */
export interface ANCClientType extends BaseClient {
    type: 'Client';
    dateCreated: string;
    serverVersion: number;
    clientApplicationVersion: number;
    clientDatabaseVersion: number;
    baseEntityId: string;
    identifiers: { [key: string]: string | null };
    addresses: FlexObject[];
    attributes: FlexObject;
    firstName: string;
    lastName: string;
    birthdate: string;
    middleName?: string;
    birthdateApprox: boolean;
    deathdateApprox: boolean;
    gender?: string;
    relationships: {
        [key: string]: string[];
    };
    id: string;
    revision: string;
}

/** reducer name for the ANC module */
const reducerName = 'ANC';

/** ANC Reducer */
const reducer = reducerFactory<ANCClientType>(reducerName);

// action
/** actionCreator returns action to to add anc records to store */
const fetchANC = fetchClientsFactory<ANCClientType>(reducerName);

// selectors
const getAllANCById = getClientsByIdFactory<ANCClientType>(reducerName);
const getANCById = getClientByIdFactory<ANCClientType>(reducerName);
const getAllANCArray = getClientsArrayFactory<ANCClientType>(reducerName);
const removeANCAction = removeClientsFactory(reducerName);

export { reducerName, removeANCAction, reducer, getAllANCArray, getANCById, getAllANCById, fetchANC };
