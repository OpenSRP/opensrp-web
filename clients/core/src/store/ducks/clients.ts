import { FlexObject } from '../../helpers/utils';
import {
    BaseDux,
    fetchFactory,
    removeFactory,
    setTotalRecordsFactory,
    reducerFactory,
    getItemsByIdFactory,
    getItemsArrayFactory,
    getItemByIdFactory,
    getTotalRecordsFactory,
} from './baseDux';

/** Interface for client object as received from clientServices */
export interface Client extends BaseDux {
    dateCreated: number;
    serverVersion: number;
    clientApplicationVersion: number;
    clientDatabaseVersion: number;
    identifiers: { [key: string]: string | null };
    addresses: FlexObject[];
    attributes: FlexObject;
    firstName: string;
    lastName: string;
    birthdate: number;
    middleName?: string;
    birthdateApprox: boolean;
    deathdateApprox: boolean;
    gender?: string;
    relationships: {
        [key: string]: string[];
    };
    _id: string;
    _rev: string;
}

/** The reducer name */
export const reducerName = 'opensrp-web/client-type/clients';

/** Client Reducer */
const reducer = reducerFactory<Client>(reducerName);

// action creators
export const fetchClients = fetchFactory<Client>(reducerName);
export const removeClients = removeFactory(reducerName);
export const setTotalRecords = setTotalRecordsFactory(reducerName);

// Selectors
export const getClientsById = getItemsByIdFactory<Client>(reducerName);
export const getClientsArray = getItemsArrayFactory<Client>(reducerName);
export const getClientById = getItemByIdFactory<Client>(reducerName);
export const getTotalRecords = getTotalRecordsFactory(reducerName);

export default reducer;
