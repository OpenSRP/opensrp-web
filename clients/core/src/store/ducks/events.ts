import { keyBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'events';

/** Interface for event object as received from event search */
export interface Event {
    type: 'Event';
    dateCreated: number;
    serverVersion: number;
    clientApplicationVersion: number;
    clientDatabaseVersion: number;
    identifiers: { [key: string]: string | null };
    baseEntityId: string;
    locationId: string;
    eventDate: number;
    eventType: string;
    formSubmissionId: string;
    providerId: string;
    duration: number;
    obs: Array<{ [key: string]: string[] | string | null }>;
    entityType: string;
    version: number;
    teamId: string;
    team: string;
    _id: string;
    _rev: string;
    isSendToOpenMRS: string;
}

/** EVENTS_FETCHED action type */
export const EVENTS_FETCHED = 'opensrp/reducer/event/EVENTS_FETCHED';
/** REMOVE_EVENTS action type */
export const REMOVE_EVENTS = 'opensrp/reducer/event/REMOVE_EVENTS';

/** interface for fetch events */
export interface FetchEventsAction extends AnyAction {
    eventsById: { [key: string]: Event };
    type: typeof EVENTS_FETCHED;
}

/** interface for remove events */
export interface RemoveEventsAction extends AnyAction {
    eventsById: { [key: string]: Event };
    type: typeof REMOVE_EVENTS;
}

/** Fetch events action creator
 * @param {Event} events - events to add to store
 * @return {FetchEventsAction} - an action to add events to redux store
 */
export const fetchEvents = (eventsList: Event[]): FetchEventsAction => ({
    eventsById: keyBy(eventsList, (event: Event) => event.baseEntityId),
    type: EVENTS_FETCHED,
});

/** Fetch events action creator
 * @param {Event} events - events to add to store
 * @return {FetchEventsAction} - an action to add events to redux store
 */
export const removeEventsAction = (): RemoveEventsAction => ({
    eventsById: {},
    type: REMOVE_EVENTS,
});

/** Create type for event reducer actions */
export type EventActionTypes = FetchEventsAction | RemoveEventsAction | AnyAction;

/** interface for event state in redux store */
interface EventState {
    eventsById: { [key: string]: Event };
}

/** Create an immutable event state */
export type ImmutableEventState = EventState & SeamlessImmutable.ImmutableObject<EventState>;

/** initial events-state state */
const initialState: ImmutableEventState = SeamlessImmutable({
    eventsById: {},
});

/** the event reducer function */
export default function reducer(
    state: ImmutableEventState = initialState,
    action: EventActionTypes,
): ImmutableEventState {
    switch (action.type) {
        case EVENTS_FETCHED:
            return SeamlessImmutable({
                ...state,
                eventsById: { ...state.eventsById, ...action.eventsById },
            });
        case REMOVE_EVENTS:
            return SeamlessImmutable({
                ...state,
                eventsById: action.eventsById,
            });

        default:
            return state;
    }
}

// Selectors

/** returns an array of events
 * @param {Partial<Store>} state - the redux store
 * @return { Event } - events array
 */
export function getEventsArray(state: Partial<Store>): Event[] {
    return values(getEventsById(state));
}

/** returns the events with their ids
 * @param {Partial<Store>} state - the redux store
 * @return { Event } - event objects as values, respective ids as keys
 */
export function getEventsById(state: Partial<Store>): { [key: string]: Event } {
    return (state as any)[reducerName].eventsById;
}
