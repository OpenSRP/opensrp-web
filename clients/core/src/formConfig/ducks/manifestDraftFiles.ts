import {
    fetchActionCreatorFactory,
    removeActionCreatorFactory,
    reducerFactory,
    getItemsByIdFactory,
    getItemsArrayFactory,
    getItemByIdFactory,
} from '@opensrp/reducer-factory';
import { ManifestFilesTypes } from './manifestFiles';

/** reducer name */
export const reducerName = 'manifestDraftFiles';

/** manifest draft files Reducer */
const reducer = reducerFactory<ManifestFilesTypes>(reducerName);

// action
/** fetch manifest draft files to store action */
export const fetchManifestDraftFiles = fetchActionCreatorFactory<ManifestFilesTypes>(reducerName, 'id');

/** clear manifest draft files data in store action*/
export const removeManifestDraftFiles = removeActionCreatorFactory(reducerName);

// selectors
export const getAllManifestDraftFilesById = getItemsByIdFactory<ManifestFilesTypes>(reducerName);
export const getManifestDraftFilesById = getItemByIdFactory<ManifestFilesTypes>(reducerName);
export const getAllManifestDraftFilesArray = getItemsArrayFactory<ManifestFilesTypes>(reducerName);

export default reducer;
