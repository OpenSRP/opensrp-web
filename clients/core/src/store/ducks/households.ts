import { Client } from './clients';
import {
    fetchActionCreatorFactory,
    removeActionCreatorFactory,
    setTotalRecordsFactory,
    reducerFactory,
    getItemsByIdFactory,
    getItemsArrayFactory,
    getItemByIdFactory,
    getTotalRecordsFactory,
} from './baseDux';

/** The reducer name */
export const reducerName = 'opensrp-web/client-type/households';

/** Interface for household object same as client */
export type Household = Client;

/** Client Reducer */
const reducer = reducerFactory<Household>(reducerName);

// action creators
export const fetchHouseholds = fetchActionCreatorFactory<Household>(reducerName, 'baseEntityId');
export const removeHouseholds = removeActionCreatorFactory(reducerName);
export const setTotalRecords = setTotalRecordsFactory(reducerName);

// selectors
export const getHouseholdsById = getItemsByIdFactory<Household>(reducerName);
export const getHouseholdsArray = getItemsArrayFactory<Household>(reducerName);
export const getHouseholdById = getItemByIdFactory<Household>(reducerName);
export const getTotalRecords = getTotalRecordsFactory(reducerName);

export default reducer;
