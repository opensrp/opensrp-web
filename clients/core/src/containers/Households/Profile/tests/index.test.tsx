import { shallow, mount } from 'enzyme';
import { createLocation, createMemoryHistory } from 'history';
import * as React from 'react';
import { match, MemoryRouter } from 'react-router';
import { household1 } from '../../../../store/ducks/tests/fixtures';
import ConnectedHouseholdProfile, { HouseholdProfile } from '../index';
import * as fixtures from '../../../../../src/store/ducks/tests/fixtures';
import { fetchClients, removeClients } from '../../../../store/ducks/clients';
import { fetchHouseholds } from '../../../../store/ducks/households';
import { fetchEvents } from '../../../../store/ducks/events';
import { Provider } from 'react-redux';
import store from '../../../../store';

const history = createMemoryHistory();
const path = `/household-profile/:id/`;

const matchVariable: match<{ id: string }> = {
    isExact: false,
    params: { id: fixtures.household1.baseEntityId },
    path,
    url: path.replace(':id', fixtures.household1.baseEntityId),
};

const location = createLocation(matchVariable.url);

jest.mock('../../../../configs/env');
describe('containers/households/Profile', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let classMock: any;
    beforeEach(() => {
        jest.resetAllMocks();

        const clientListMock = jest.fn(async () => {
            return { clients: fixtures.clients, total: 7 };
        });

        const clientMock = jest.fn(async () => {
            return fixtures.households;
        });

        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        classMock = jest.fn((a, b, c) => {
            return {
                list: b === 'client/search' ? clientMock : clientListMock,
            };
        });
    });

    it('renders without crashing', () => {
        const props = {
            events: [],
            fetchClient: fetchHouseholds,
            fetchEvents: fetchEvents,
            fetchMembers: fetchClients,
            removeMembers: removeClients,
            history,
            household: household1,
            location,
            match: matchVariable,
            members: [],
            opensrpService: classMock,
        };
        const component = shallow(<HouseholdProfile {...props} />);
        const wrapper = component.find('#household-profile');
        expect(wrapper.length).toBe(1);
    });

    it('render the basic information and member list correctly', async () => {
        const props = {
            events: [],
            fetchClient: fetchHouseholds,
            fetchEvents: fetchEvents,
            fetchMembers: fetchClients,
            removeMembers: removeClients,
            history,
            household: household1,
            location,
            match: matchVariable,
            members: [],
            opensrpService: classMock,
        };
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter initialEntries={['/']} keyLength={0}>
                    <ConnectedHouseholdProfile {...props} />
                </MemoryRouter>
            </Provider>,
        );

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        await new Promise((resolve: any) => setImmediate(resolve));
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
    });
});
