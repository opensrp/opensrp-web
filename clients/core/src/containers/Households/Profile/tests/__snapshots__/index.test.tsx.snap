// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/households/Profile render the basic information and member list correctly 1`] = `ReactWrapper {}`;
